Introduced owned Difference

B-2U opened the pull request:
Here's a prototype of my approach, 
The main thing I want is to keep the original macros and the no_panic function as fast as before.

When I was trying to make a `DifferenceBuf` like PathBuf and String in std, I realized they are struct but not a memory window like &str, so I can never create a `Difference<'_>`  cheaply from it

So I take `DifferenceBuf` as a public variant of `Difference<'_>`, it's used only for the case when we want to return a struct to crate user, and since it's the only struct user can see (`Difference<'_>` is still private), I rename `Difference<'_>` to `DifferenceRef<'_>`, and let `DifferenceBuf`-> `Difference`

The disadvantage is that I probably added too much code to achieve it and renamed the original `assert_json_matches_no_panic()` to `assert_json_matches_no_panic_to_string()`, which makes it a breaking change. I can't really figure out a good name for the new function.

Please let me know what you think about it. Thanks!

B-2U committed:
update actions in workflow to get rid of deprecated warnings

hardselius commented on src/lib.rs:
````diff
@@ -309,6 +311,69 @@ where
+/// let result = assert_json_matches_no_panic(&lhs, &rhs, &config);
+/// assert!(result.is_err());
+/// ```
+pub fn assert_json_matches_no_panic<Lhs, Rhs>(
````
Maybe we should drop the `assert_` prefix altogether? This isn't really an assert, because in my book an assert is supposed to panic. And giving it a new name isn't breaking. So maybe

```suggestion
pub fn json_matches<Lhs, Rhs>(
```

is better? And if you make `Difference` `Debug + Display`, it's a proper `Error`. I think this would make the old `assert_json_matches_no_panic` redundant, right?

hardselius replied:
Or maybe something like `check_json_matches` or `test_json_matches` to convey that it's something that could fail, and not something that returns a boolean.

hardselius replied:
I'm discussing with myself here, but after some thought I think I've landed on a good name.

`try_assert_json_matches`

I think the `try_assert_` prefix is compelling because it aligns well with Rust's existing conventions and indicates that it is a fallible version of the `assert_` macro.

B-2U replied:
`try_assert_json_matches` sound good!

> I think this would make the old assert_json_matches_no_panic redundant, right?

The reason I wanted to keep it is for the speed, since the assert macros and the old `assert_json_matches_no_panic` itself are still using the borrow type `DifferenceRef<'_>` which should be faster than the new owned one.

B-2U replied:
I just made a quick experiment, to figure out the overhead if we completely moved to the new `Difference` (So it like the Option 1 mentioned in https://github.com/hardselius/serde-json-assert/issues/1 )
test branch: https://github.com/B-2U/serde-json-assert/commit/1f5b781119a4473f05315837fdba3ee467089b79

When I ran `cargo test` on my pc (Intel i7 10700), all normal tests finished in 0.00 or 0.01s, the same as the original main branch, 
and for the slowest doc tests, the main branch took about 0.28 ~ 0.32s, and this option 1 branch took 0.29 ~ 0.35s, about 9% overhead in both worst case. 

In my use cases, I do not really care about the speed of json assertion, but maybe it will be useful for some other users?

B-2U replied:
Another advantage of let the macros use the new method (Option 1) is it will be easier to impl for https://github.com/hardselius/serde-json-assert/issues/2
since the macros will get a real struct but not string

B-2U replied:
I just make another [test](https://github.com/B-2U/serde-json-assert/commit/97a58b11233007372be690334dcac8c215da135a) that run comparison 100k times for **Option 1** on the branches `main` vs `make-diff-owned` ( directly making `Difference<'_>` a owned type ) 
The overhead of directly making `Difference<'_>` in main branch to own its data is around 5%

![image](https://github.com/user-attachments/assets/770c34b9-f79e-4ef4-b021-a8a2136a838f)
![image](https://github.com/user-attachments/assets/fcacc9f3-a971-48eb-bb55-b518f1b295e6)

Only 5% overhead is kind of attractive to me, since we wouldn't need 3 more new types and boilerplates ( approach in this PR ), and it will be more friendly for #2.

Please let me know what you think, is option 1 a better idea for you?

hardselius reviewed (changes requested):
See comment and let me know what you think.

Also, make sure to squash/rebase you commits into just two, one for the change and one for the workflow, at the end of this.

hardselius commented:
I also encourage you to add a sentence or two to the `CHANGELOG.md`.

hardselius commented:
I think it should be fine to pay the performance cost in most cases. The best way to do this is probably to revert the name change of the `assert_json_matches_no_panic` function and deprecate it. Let give users a minor version or two to migrate without immediately breaking their builds.

But then again, if this is all to use inputs that are `Serialize`, should we bend over backwards for that? Maybe the extra code is better for now. We can always consolidate later. Expose the new types, using the existing names, and rename the existing types by adding the `Ref` postfix.

If #2 requires the owned type, we can consolidate sooner. I think this gives us the most flexibility going forward.

B-2U commented:
> But then again, if this is all to use inputs that are `Serialize`, should we bend over backwards for that? Maybe the extra code is better for now. We can always consolidate later. Expose the new types, using the existing names, and rename the existing types by adding the `Ref` postfix.
> 
> If #2 requires the owned type, we can consolidate sooner. I think this gives us the most flexibility going forward.

I'm not really sure if I understand it correctly, so you suggested to 
1. Mark a`ssert_json_matches_no_panic` as deprecated and use the new `try_assert_json_matches` in assertion macros
2. Use and expose the new types with existing names,  and rename the old types with Ref prefix, so we can turn to **option 2** (take `Value` as inputs) much more easier if needed?

Please let me know if I misunderstood

hardselius commented on src/lib.rs:
```diff
@@ -262,18 +264,18 @@ macro_rules! assert_json_eq {
 ///
-/// Instead it returns a `Result` where the error is the message that would be passed to `panic!`.
+/// Instead it returns a `Result` where the error is the message string that would be passed to `panic!`.
 /// This is might be useful if you want to control how failures are reported and don't want to deal
 /// with panics.
-pub fn assert_json_matches_no_panic<Lhs, Rhs>(
+pub fn assert_json_matches_no_panic_to_string<Lhs, Rhs>(
```
Revert these changes.

hardselius commented:
Ah, sorry for being unclear.

> Use and expose the new types with existing names, and rename the old types with Ref prefix, so we can turn to option 2 (take Value as inputs) much more easier if needed?

I think we only do this at this time, and leave everything else untouched. Only expose the new `try_assert_json_matches` and the new types. That means we're halfway towards Option 1, without having painted ourselves into a corner. I think this is basically what your current changes reflect. Only revert `assert_json_matches_no_panic_to_string` to `assert_json_matches_no_panic` and rebase your commits. Does that work for you?

B-2U committed:
introduce a expose a own type Difference, rename the existing one to DifferenceRef
new pub function try_assert_json_matches which return Result<(), Vec<Difference>>

B-2U commented:
> Ah, sorry for being unclear.
> 
> > Use and expose the new types with existing names, and rename the old types with Ref prefix, so we can turn to option 2 (take Value as inputs) much more easier if needed?
> 
> I think we only do this at this time, and leave everything else untouched. Only expose the new `try_assert_json_matches` and the new types. That means we're halfway towards Option 1, without having painted ourselves into a corner. I think this is basically what your current changes reflect. Only revert `assert_json_matches_no_panic_to_string` to `assert_json_matches_no_panic` and rebase your commits. Does that work for you?

Got it! Rebased and changelog are done. 
Please take a look :D
